File: src/main/java/jv/collection/queues/QueueLinkedList.java
package jv.collection.queues;

public class QueueLinkedList<T> implements Queue<T> {
    private Node first = null;
    private Node last = null;

    @Override
    public void enqueue(T item) {
        Node node = new Node(item, null);
        if (isEmpty()) {
            first = node;
            last = node;
        } else {
            last.next = node;
            last = node;
        }
    }

    @Override
    public T dequeue() {
        if (isEmpty()) {
            throw new RuntimeException();
        }

        T value = first.value;
        first = first.next;
        return value;
    }

    @Override
    public boolean isEmpty() {
        return first == null;
    }

    private class Node {
        private final T value;
        private Node next;

        private Node(T value, Node next) {
            this.value = value;
            this.next = next;
        }
    }
}

File: src/main/scala/sc/collection/queues/QueueLinkedList.scala
package sc.collection.queues


object QueueLinkedList {

  private class Node[T](val value: T) {
    var next: Node[T] = null
  }

}

class QueueLinkedList[T] extends Queue[T] {

  import QueueLinkedList.Node

  private var first: Node[T] = null
  private var last: Node[T] = null

  def enqueue(item: T) {
    val node = new Node[T](item)
    if (isEmpty) {
      first = node
      last = node
    }
    else {
      last.next = node
      last = node
    }
  }

  def dequeue(): T = {
    if (isEmpty) throw new RuntimeException

    val value = first.value
    first = first.next

    if (isEmpty) last = null

    value
  }

  def isEmpty: Boolean = first == null
}
